### PR TITLE
Overset tut

### PR DESCRIPTION
**Pull request by DavidAnderegg**
## Purpose
This extends the MACH-Tutorial with a section for overset meshes. Additionally, geometry generation with OpenVSP and surface mesh generation with Pointwise is demonstrated.

## Type of change
What types of change is it?
- New feature (non-breaking change which adds functionality)

## Checklist
_Put an `x` in the boxes that apply._

- [ ] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

### CODE REVIEW

**Comment by DavidAnderegg** [comment]
I believe, this is ready to get some feedback. There are still two things missing:
- some reference-stuff doesnt work properly
- I want to add one more picture at the start of the surface-mesh generation

I will fix those two things in the new year and would highly appreciate, if somebody could read it and tell me where i could improve.

**Comment by DavidAnderegg** [comment]
I fixed the the reference-stuff and decided, the picture i was planning to add,  is not needed.

File: machAeroTutorials/overset_theory.rst
Comment (marcomangano) [inline machAeroTutorials/overset_theory.rst:32]: I would be a bit more explicit saying `It is recommended to avoid such overlapping near critical regions of the flowfield, like the wing tip`. Also, maybe change `calculated solution` into `numerical solution`?

File: machAeroTutorials/overset_theory.rst
Comment (marcomangano) [inline machAeroTutorials/overset_theory.rst:73]: This section could just be called `Useful tips` ;)

File: machAeroTutorials/overset_theory.rst
Comment (marcomangano) [inline machAeroTutorials/overset_theory.rst:103]: This section looks super helpful! Could you just briefly mention /  add links to better understand the flooding and orphan cells concepts? Just a couple of sentences should do the job.

File: machAeroTutorials/overset_surface_mesh.rst
Comment (marcomangano) [inline machAeroTutorials/overset_surface_mesh.rst:37]: I find this figure not super clear, but I understand the challenge of visualizing the three separate meshes in 3D. 
Maybe you can add a 3D view of the wing with just the surface mesh (highlighting the different near field meshes) and a front/side 2D view to show the overlapping?

File: machAeroTutorials/overset_volume_mesh.rst
Comment (marcomangano) [inline machAeroTutorials/overset_volume_mesh.rst:68]: Can you add a link to `pyhyp` options documentation for this and the following section? Makes it easier to jump between different RTD docs.

File: machAeroTutorials/overset_analysis.rst
Comment (marcomangano) [inline machAeroTutorials/overset_analysis.rst:163]: missing space between words

**Review by marcomangano — COMMENTED** [review]
Hi David, thanks a lot for putting this big effort into this tutorial, it really looks good and it will definitely be useful.
I did not check the `openVSP` and `Pointwise` tutorials because I don't have the software, but I gave an overall look at the other sections and added a few comments for typos and minor additions (see below).
As additional comment, I would just recommend to not use the first person in the explanations, for consistency with the rest of the tutorials. Your contribution will still be evident via the `about` section in the main page, the links to `adflow_utils`, and the git log history!

Let me know if you want me and other students to check / try a section more in detail. I guess we need people to start using the tutorial extensively to catch all the fine details we can add or improve. This is already pretty close to be merged imho. Thanks again!

File: machAeroTutorials/overset_theory.rst
Comment (DavidAnderegg) [reply machAeroTutorials/overset_theory.rst:13]: Hi Marco, thanks for taking a look. I agree with all of your points and will update them.

I will also change the first person stuff. I just wrote how it came to my mind withouth thinking about this.

It would be great if you somebody could work through it as a test. OpenVSP is completley free. I believe there is a trial version of pointwise that might be handy.

File: machAeroTutorials/overset_theory.rst
Comment (DavidAnderegg) [reply machAeroTutorials/overset_theory.rst:103]: I will try. I just copied a presentation about this, that Anil gave me. I never really bothered to understand it.

Comment (joanibal) [reply machAeroTutorials/overset_theory.rst:103]: https://mdolab.engin.umich.edu/bibliography/Kenway2017a.html

Here is the paper that talks about the details

Comment (joanibal) [reply machAeroTutorials/overset_theory.rst:103]: Never mind, I see that you have already done so.

File: machAeroTutorials/overset_volume_mesh.rst
Comment (joanibal) [inline machAeroTutorials/overset_volume_mesh.rst:169]: It would be great to have a picture here of the final mesh

File: machAeroTutorials/overset_analysis.rst
Comment (joanibal) [inline machAeroTutorials/overset_analysis.rst:204]: :+1:  Great addition

**Review by joanibal — APPROVED** [review]
Thank you for this great tutorial. 
It would be nice to have a picture of the final mesh, but not a requirement.

**Comment by anilyil** [comment]
Thank you for the great work @DavidAnderegg! I have a few comments:
1. I see the overset theory part is mostly based on the pdf document by Ney Secco that I sent you. Does this tutorial cover all of the material in that pdf, or is there still value in referring users to that document also? If this guide covers all of the materials in there, then it is good as is.
2. I want to rephrase some minor items that I saw in the document. Do you mind if I edit these directly and push commits myself? I think as a reviewer to this PR, I can push commits to your branch. Please let me know if this is okay.

Besides these, I am mostly okay with approving this PR. Let me know what you think about these two items.

**Comment by anilyil** [comment]
@nwu63 reminded me that we also have a text file with useful tips. I am attaching the file to this comment. Can we add these in the theory section @DavidAnderegg? I think just copy pasting these should be fine for now. Do you think there is any overlap with the material you already present in there now?

The PDF you incorporated in the tutorial and this text file is from a previous student from the lab, Ney Secco. I think we should also acknowledge the materials we are re-using from his work. I can make this modification though.

[overset_tips.txt](https://github.com/mdolab/MACH-Aero/files/5832377/overset_tips.txt)

**Comment by DavidAnderegg** [comment]
@anilyil, I am happy to hear that you like it.
1. I dont think there is value in referring users to the pdf, but the txt file you attached has a bit more extended information than i have written. I think this is a question of what one wants. When i started figuring out how ADflow implemented overset-meshes i had absolutely no clue about this concept. I wrote the tutorial with that in mind. My explanations are maybe not 100% accurate and not as in-depth as they could be, but i believe they are easy to understand. It might make sense to link/copy the txt-file. We should absolutely give him credit.
2. Yes, I am fine with that. Let me know if you dont have access.

**Comment by anilyil** [comment]
All sounds good. I will make the changes and push to this branch.

**Comment by anilyil** [comment]
@nwu63 this PR has a few relatively large binary files. In particular, see `tutorial/overset/mesh/onera_m6.pw`, which is about 5mb. I think this is useful as a guide to people who want to use Pointwise, but I am not sure where we stand on how much storage we can use in repos like this. 

Is including these files okay in this PR? Should we try to reduce the size of the files added with this PR?

**Comment by nwu63** [comment]
Yes good point. I'm not sure we have a policy in place for that, for example there are existing CGNS meshes in this repo and I'm not overly thrilled about those either. I think these should probably be stored elsewhere as a zip file of supplementary material, since it's not essential for the progression of the tutorial (I think) whereas the CGNS files are.

However, it's notoriously difficult to purge commits in git so even if the commit gets reverted or deleted via rebasing, the file is still there somewhere and will continue to bloat the size of the repo. There are ways to clean it up but it's somewhat convoluted, so maybe this is fine for now until we figure out a consistent strategy.

File: machAeroTutorials/overset_volume_mesh.rst
Comment (anilyil) [reply machAeroTutorials/overset_volume_mesh.rst:169]: @joanibal I see that this comment is still not addressed. Do you want to have this figure? It is kinda difficult to show a figure of the overset volume mesh so I am not sure how it would look like. Let me know if you still want this figure.

**Comment by anilyil** [comment]
@DavidAnderegg I pushed a few commits and updated the tutorial in a few places. I mainly modified the theory part. Besides that, my changes are mostly cosmetic. I also compressed a few large figures. Please let me know if these changes look okay to you.

**Comment by DavidAnderegg** [comment]
@anilyil all changes look okey. Thank you.

**Comment by anilyil** [comment]
Hello all, I finally managed to get back to this. Ney had a few more figures in his pdf, and I updated the overset theory section to include all of the material he had basically. I think some of these will be superseded by the collar mesh tutorial from @sseraj but something is better than nothing until then. 

@sseraj and @joanibal can you please review the overset theory section again? Feel free to reword stuff as you see fit. My goal was to just copy paste as much as Ney's stuff directly in there, so sorry if its not organized.

Once everyone is happy with this PR, we should take a look at the file sizes. I think there are large files in the history here, so we may need to do a similar thing as we are doing with the PR: https://github.com/mdolab/pygeo/pull/72 @nwu63 I will need to coordinate with you when we get to that point.

**Comment by anilyil** [comment]
Also we can convert png figures to jpg to save more space. I already did that for the largest png files but there is still a lot more compression we can do to reduce the file sizes.

**Comment by anilyil** [comment]
Also converting this to a draft to prevent an accidental merge. We will take care of the large files in history before we merge this.

File: machAeroTutorials/overset_theory.rst
Comment (marcomangano) [inline machAeroTutorials/overset_theory.rst:95]: his looks super relevant, do you think we should turn this into a note/warning or is it "common knowledge"?

File: machAeroTutorials/overset_theory.rst
Comment (marcomangano) [inline machAeroTutorials/overset_theory.rst:150]: I guess the second sentence is another bullet (which also explains the missing character there?)

File: machAeroTutorials/overset_theory.rst
Comment (marcomangano) [inline machAeroTutorials/overset_theory.rst:236]: This list is not working as expected?

File: machAeroTutorials/overset_theory.rst
Comment (marcomangano) [inline machAeroTutorials/overset_theory.rst:238]: The file is in the private tutorial repo

**Review by marcomangano — COMMENTED** [review]
Great addition! check my formatting comments below

File: machAeroTutorials/overset_theory.rst
Comment (anilyil) [reply machAeroTutorials/overset_theory.rst:95]: Converted this to a note.

File: machAeroTutorials/overset_theory.rst
Comment (anilyil) [reply machAeroTutorials/overset_theory.rst:150]: Yes! fixed

File: machAeroTutorials/overset_theory.rst
Comment (anilyil) [reply machAeroTutorials/overset_theory.rst:236]: It was not working; fixed it.

File: machAeroTutorials/overset_theory.rst
Comment (anilyil) [reply machAeroTutorials/overset_theory.rst:238]: What do you suggest we do about this? Is it worth copying over to this repo as well? Should I say this is available in the internal tutorials or sth like that?

**Comment by anilyil** [comment]
Thank you for the quick review @marcomangano! I addressed your comments except for the last one. Let me know what you think about that. Any ideas on that last item @nwu63?

**Comment by anilyil** [comment]
Also, it looks like @nwu63 and @bernardopacini tested squash merges with large files in the history and found that as long as the PR is from a separate fork, these files in the history do not show up in the history of this PR. So once this is approved, I think we can merge as is and we won't need to deal with the large files in the history.

File: machAeroTutorials/overset_theory.rst
Comment (DavidAnderegg) [reply machAeroTutorials/overset_theory.rst]: I would prefer if this file is publicly available if you mention it in the tutorial.

Maybe it would make sense to drop it to the ADflow repo under examples...

**Comment by DavidAnderegg** [comment]
@anilyil I found a wrong caption and tried to mark it in my commit (I dont know how i can target my comment to a particular file and line number). I also think it makes sense to explain that adflow only writes the iBlank if you tell it to do so. (This confused me at the beginning for quite a while...)

**Comment by anilyil** [comment]
Good catch with the caption, I will fix it. Thanks! Also thanks for adding the iBlank output statement.

**Comment by anilyil** [comment]
I fixed the caption for the stencil, and also updated the figure. The previous figure was used in the stencil coloring figure for Peter's paper, and I thought the colors used in that figure directly might confuse people with the iBlank contours. So I just put in the simplified stencil figure from the ank paper.

**Comment by marcomangano** [comment]
> Thank you for the quick review @marcomangano! I addressed your comments except for the last one. Let me know what you think about that. Any ideas on that last item @nwu63?

Agreed, I think that if we mention the file we should make it available among the other tutorial examples. Is there a reason why it should not be publicly available? Do we want to make it "available upon request" (probably useless and more time consuming for us)? calling in @sseraj @eirikurj for additional feedback

**Review by marcomangano — CHANGES_REQUESTED** [review]
I just realized that the example files are not linted and formatted correctly. Can you please fix them real quick so all the checks pass?
I am referring to the python scripts in `tutorial/overset/`

**Comment by anilyil** [comment]
ran formatter, lets see if they pass now

**Comment by marcomangano** [comment]
> ran formatter, lets see if they pass now

Just [a few manual edits](https://dev.azure.com/mdolab/Public/_build/results?buildId=780&view=logs&jobId=72ffc5b4-e866-5000-bf94-31e63633fac7&j=a5dba9ca-28d5-5c60-b696-bd0c6902acd4&t=780fb814-8e0b-5c54-9ada-dece6ee78af7) left :)

**Comment by marcomangano** [comment]
L M A O
```
2021-03-19T12:07:23.3197081Z ./tutorial/overset/analysis/plot_grid_convergence.py:113:1: W391 blank line at end of file
2021-03-19T12:07:23.3198476Z 
2021-03-19T12:07:23.3199651Z ^
2021-03-19T12:07:23.3200142Z 1     W391 blank line at end of file
```

**Review by marcomangano — APPROVED** [review]
Success! Let's just wait to see if we want to add the missing file

File: machAeroTutorials/overset_theory.rst
Comment (nwu63) [reply machAeroTutorials/overset_theory.rst]: From a cursory glance, it seems to just set `surfacevariables` and `volumevariables` to `['blank']` and then call `writeVolumeSolutionFile()`. So maybe we can just describe the steps here? Alternatively we can duplicate the file here and add it to the repo.

Comment (anilyil) [reply machAeroTutorials/overset_theory.rst]: actually yeah, so I think we can just add the file here? Do you want to avoid code duplication? If so, I can also copy the commands as a code block here.

Comment (nwu63) [reply machAeroTutorials/overset_theory.rst]: It looks like the script was set up specifically for the case in the private repo. If the script can be generalized, the best approach would be as @DavidAnderegg mentioned, i.e. convert it to a general script and put it in ADflow. Do you think that's doable?

Comment (anilyil) [reply machAeroTutorials/overset_theory.rst]: In the interest of not delaying this PR any more, I just added the ihc_check script here and added some information about it in the tutorial docs. Let me know how it looks.

Comment (nwu63) [reply machAeroTutorials/overset_theory.rst]: Looks fine, I'll open an issue to move it over to adflow later.

**Review by nwu63 — DISMISSED** [review]

File: machAeroTutorials/overset_theory.rst
Comment (anilyil) [reply machAeroTutorials/overset_theory.rst]: makes sense, thanks!

**Comment by anilyil** [comment]
@DavidAnderegg I tried running the script and I thought they worked but they did not run. The mesh generation gets negative volumes on the tip section. The hole cutting also failed for me for both L2 and L3. Did not try L1 yet. Were the cases working for you when you created the PR?

**Comment by DavidAnderegg** [comment]
@anilyil Strange. It did work. 

Let me check, may be i switched something up?!?

**Comment by anilyil** [comment]
I have a change in the volume mesh extrusion to fix the issue with the current surface meshes. But the repo as is did not run. I also realized I needed to update some adflow options for capitalization

**Review by sseraj — COMMENTED** [review]
I agree with not delaying this PR, but I would like to make several changes to the overset theory. In particular, I would prefer to have Ney's text file be integrated into the rest of the page and remove duplicate and outdated information. Would it be better if I make an issue and address this in a separate PR?

**Comment by anilyil** [comment]
> Would it be better if I make an issue and address this in a separate PR?

I think you know what my preference will be there :)

**Comment by anilyil** [comment]
Also the overset theory now has a whole bunch of debugging help etc. We can reorganize that in a separate PR imo, we can also address your point.

**Comment by DavidAnderegg** [comment]
@anilyil I just tried aswell and it also failed. As the capitalization indicates, it was written for an older version of MACH-Aero. But I highly doubt this affected this. So i have actually no clue how that happend...

I also just checked the log files when i ran the simulations. ADflow does usually give a warning if there are negative volumes in the mesh, right? There are no warnings in the log I have.

On a general note, You dont have to hurry with this PR because of me. I never saw it as a high priority thing and I prefer it beeing helpfull an complete.

**Comment by sseraj** [comment]
> Also the overset theory now has a whole bunch of debugging help etc. We can reorganize that in a separate PR imo, we can also address your point.

I agree. I created issue #41. Feel free to add other to dos.

**Comment by anilyil** [comment]
Yeah, I just want to get this done because it has been on my list for a while. Do you remember updating the surface mesh files at any point? I can update the runscript options but I doubt we made major changes to the math while you were working on this.

**Comment by DavidAnderegg** [comment]
I am kicking into investigation mode. I just compared the output surface-meshes to the ones that lie in the repo. They are exactly the same.

It just came to my mind that when pyhyp was updated, some variables that could have been integers, now have to be float. Might that be the reason? I kinda doubt it, but that's the best i got at the moment...

**Comment by anilyil** [comment]
@DavidAnderegg there is a block missing in the wing-tip mesh. its the one right near the tip on the TE. I could not see it in the volume extrusion. Its also not there on the surface mesh. You probably updated the surface mesh files at some point?

**Comment by DavidAnderegg** [comment]
You are right. I did somehow not export this part, but in my analysis-files it is not missing. Strange. I will commit the proper surface-file. 

But I also just noticed ``simpleOCart`` does apparently fail because the pyhyp options were changed (farfield must be lowercase now)

**Comment by anilyil** [comment]
oh neat, I will update the pyhyp options and try again. the L3 mesh in the analysis folder did work with the overset

**Comment by sseraj** [comment]
> But I also just noticed `simpleOCart` does apparently fail because the pyhyp options were changed (farfield must be lowercase now)

Is this with an old version of cgnsUtilities? I think I fixed this at the same time as the pyHyp update

**Comment by DavidAnderegg** [comment]
> > But I also just noticed `simpleOCart` does apparently fail because the pyhyp options were changed (farfield must be lowercase now)
> 
> Is this with an old version of cgnsUtilities? I think I fixed this at the same time as the pyHyp update

This might very well be the case... I updated approx 2 months ago

**Comment by anilyil** [comment]
I just updated all repos locally and things work. Will push an update in a second.

**Comment by anilyil** [comment]
The meshes run and hole cutting script works now. I have created an issue on the mesh parameters but thats a secondary thing in my opinion. The issue is up at #42.

@joanibal do you want to review this? I dont want to put an extra burden on you. I think this is probably good to go for now but if you want to, feel free to take a look and leave comments.

**Review by joanibal — APPROVED** [review]

**Comment by nwu63** [comment]
Thanks everyone and especially @DavidAnderegg for your contributions! :tada: :clap:

**Pull request merged**
